modules/renter/download.go

```diff
@@ -1,805 +1,484 @@
+		priority      uint64 // Files with a higher priority will be downloaded first.
 	}
 )
 
-// newSectionDownload initializes and returns a download object for the specified chunk.
-func (r *Renter) newSectionDownload(f *file, destination modules.DownloadWriter, offset, length uint64) *download {
-	d := newDownload(f, destination)
-
-	if length == 0 {
-		build.Critical("download length should not be zero")
-		d.fail(errors.New("download length should not be zero"))
-		return d
-	}
-
-	// Settings specific to a chunk download.
-	d.offset = offset
-	d.length = length
-
-	// Calculate chunks to download.
-	minChunk := offset / f.chunkSize()
-	maxChunk := (offset + length - 1) / f.chunkSize() // maxChunk is 1-indexed
-
-	// mark the chunks as not being downloaded yet
-	for i := minChunk; i <= maxChunk; i++ {
-		d.finishedChunks[i] = false
-	}
-
-	d.initPieceSet(f, r)
-	return d
-}
-
-// newDownload creates a newly initialized download.
-func newDownload(f *file, destination modules.DownloadWriter) *download {
-	return &download{
-		startTime:        time.Now(),
-		chunkSize:        f.chunkSize(),
-		destination:      destination,
-		erasureCode:      f.erasureCode,
-		fileSize:         f.size,
-		masterKey:        f.masterKey,
-		numChunks:        f.numChunks(),
-		siapath:          f.name,
-		downloadFinished: make(chan struct{}),
-		finishedChunks:   make(map[uint64]bool),
-	}
-}
-
-// initPieceSet initializes the piece set, including calculations of the total download size.
-func (d *download) initPieceSet(f *file, r *Renter) {
-	// Allocate the piece size and progress bar so that the download will
-	// finish at exactly 100%. Due to rounding error and padding, there is not
-	// a strict mapping between 'progress' and 'bytes downloaded' - it is
-	// actually necessary to download more bytes than the size of the file.
-	// The effective size of the download is determined by the number of chunks
-	// to be downloaded. TODO: Handle variable-size last chunk - Same in downloadqueue.go
-	numChunks := uint64(len(d.finishedChunks))
-
-	dlSize := d.length
-	d.reportedPieceSize = dlSize / (numChunks * uint64(d.erasureCode.MinPieces()))
-	d.atomicDataReceived = dlSize - (d.reportedPieceSize * numChunks * uint64(d.erasureCode.MinPieces()))
-
-	// Assemble the piece set for the download.
-	d.pieceSet = make(map[uint64]map[types.FileContractID]pieceData)
-	for i := range d.finishedChunks {
-		d.pieceSet[i] = make(map[types.FileContractID]pieceData)
-	}
-
-	f.mu.RLock()
-	for _, contract := range f.contracts {
-		id := r.hostContractor.ResolveID(contract.ID)
-		for i := range contract.Pieces {
-			// Only add pieceSet entries for chunks that are going to be downloaded.
-			m, exists := d.pieceSet[contract.Pieces[i].Chunk]
-			if exists {
-				m[id] = contract.Pieces[i]
-			}
-		}
+// staticComplete is a helper function to indicate whether or not the download
+// has completed.
+func (d *download) staticComplete() bool {
```

Comment:
static is a new convention that indicates you can access the field holding or not holding a lock.

if it were `complete`, it would be implied that a lock is required. If it were `managedComplete`, it would be implied that you must **not** be holding a lock. But static implies that it doesn't matter. Seems weird for a function, but it's true because the implementation of the function uses channels.